src/persist-client/src/read.rs

```diff
@@ -0,0 +1,137 @@
+        todo!("{:?}", cx)
+    }
+}
+
+pub struct Tail<K, V, T, D>(PhantomData<(K, V, T, D)>);
+
+impl<K, V, T, D> Stream for Tail<K, V, T, D> {
+    // WIP: We also need a way to delivery progress information from this.
+    type Item = (K, V, T, D);
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        todo!("{:?}", cx)
+    }
+}
+
+pub struct ReadHandle<K, V, T, D> {
+    _phantom: PhantomData<(K, V, T, D)>,
+}
+
+impl<K, V, T, D> ReadHandle<K, V, T, D>
+where
+    K: Codec,
+    V: Codec,
+    T: Timestamp + Codec64,
+    D: Semigroup + Codec64,
+{
+    // This handle's since frontier, not the global collection-level one.
+    pub fn since(&self) -> &Antichain<T> {
+        todo!()
+    }
+
+    // NB new_upper of the empty antichain "finishes" this collection, promising
+    // that no more data will ever be read by this handle.
+    //
+    // WIP AntichainRef?
+    pub fn downgrade_since(&self, new_since: Antichain<T>) {
+        todo!("{:?}", new_since)
+    }
+
+    // WIP: The impl is far far easier if this is something we keep single
+    // process at first (as opposed to being able to shard it like snapshots).
+    // Do we think that's fine for short term (i.e. initial platform launch)? We
+    // can definitely revisit this later if it ends up being a bottleneck.
+    //
+    // WIP: This is an opportunity to push down Projection information to save
+    // bandwidth. My thinking so far is that Map and Filter should stay
+    // STORAGE-level concerns.
+    //
+    // WIP: My first attempt here was to return `impl Stream` but that didn't
+    // compile with `todo!`. I don't yet have an opinion on which is better.
+    pub async fn tail(as_of: Antichain<T>) -> Result<Tail<K, V, T, D>, Error> {
+        todo!("{:?}", as_of);
+    }
+
+    // NB the len of the returned vec is num_shards
+    //
+    // WIP This is an opportunity to push down Projection information to save
+    // bandwidth. My thinking so far is that Map and Filter should stay
+    // STORAGE-level concerns.
+    pub async fn snapshot(
+        &self,
+        as_of: Antichain<T>,
+        num_shards: NonZeroUsize,
+    ) -> Result<Vec<SnapshotShard>, Error> {
+        todo!("{:?}{:?}", as_of, num_shards);
+    }
+
+    // WIP: Should this live on ReadHandle or somewhere else? E.g. I think it
+    // could also live on Client.
+    //
+    // WIP: Figure out the error story here. Retries should move inside this, but
+    // what happens if we get data that we can't decode?
+    //
+    // WIP: My first attempt here was to return `impl Stream` but that didn't
+    // compile with `todo!`. I don't yet have an opinion on which is better.
+    pub async fn snapshot_shard(
+        &self,
+        shard: SnapshotShard,
+    ) -> Result<SnapshotIter<K, V, T, D>, Error> {
+        todo!("{:?}", shard);
+    }
+
+    pub async fn clone(&self) -> Result<Self, Error> {
```

Comment:
is cloning semantically equivalent to someone opening another read capability on the same collection (assuming no read capabilities have called downgrade_since) or is it different? I understand cloning in the sense of "I want to give someone a copy of exactly the same read capability that I have, with visibility through the same since". 

I guess what I'm not sure about is: if someone makes a new capability, shouldn't it be from a since <= my read capability's since? (assuming single dimensional times). And so, given all of that, its not clear to me why someone would want to clone a read cap.

Reply:
yeah, this is intended to be cloning as defined in the formalism doc, where my understanding is that a cloned read handle gets a since that is the same as this one at cloning time but independent thereafter (ditto write handle clone and upper)

I also am in the dark on when a handle will be cloned vs obtained via something like open_collection. the formalism doc really just defines things, not how they'll be used. maybe Petros has a better sense for this?

> if someone makes a new capability, shouldn't it be from a since <= my read capability's since?

I think it would have to be >=, no? sort of beside the point, since we choose =, but I want to make sure I'm not missing something

Reply:
I'm not sure what Frank had mind but maybe it's useful for active-replication when we restart instances?

Reply:
Here's my understanding. Let's say we have two read capabilities A and B for the same collection. A has a since of 5, and B has a since of 10. My understanding is that the collection's since frontier is 5, because it has to be the min of all of the since frontiers for all of the read capabilities (in single dimensional times), because if the since frontier for the collection was say 7, that seems logically inconsistent with A's since frontier of 5. This might not be how the formalism doc thinks about this -- in which case my apologies.

In this framing, a new read capability would be instantiated at the collection's since frontier (in this case 5), which is <= the since frontier of all of the existing read capabilities (A has 5, B has 10)

Reply:
oh I see. my understanding from the formalism doc is that you get the since/upper of the handle you clone, not the overall since/upper of the collection at the time you clone. I was just matching that behavior here, but we can do whatever

Reply:
yeah sorry - on clone you get the since frontier of the cap you cloned, but if you mint a new cap you get the since frontier of the collection which is <= the since frontiers of all of the existing read caps, which makes it less clear to me why you would want/need to clone

Reply:
aha! now i get you. yeah I'm curious what @petrosagg  and @aljoscha  think here on when each might be used and whether we need both